### ISSUE TITLE
Allow importing from zip over existing local instance / differential instance import

### ISSUE THREAD

**Issue by Khitiara** [MultiMC/MultiMC5#3918]
### Role

Playing modded multiplayer with friends

### Suggestion

Allow importing a zip over an existing instance with new files to overwrite and a metadata list of removed files, in similar format to standard mmc import from zip.

### Benefit

Would allow distributing pack updates to at least private packs without concern of users not being able to manage getting user-side configs and settings copied over.

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

**Comment by Xetaxheb** [comment]
how would this be any less work for the person creating the diff than a simple batch file?
putting aside the complexities of the setup... there are mods like defaultoptions for this re: configs

**Comment by Khitiara** [comment]
In my experience a batch file usually isnt actually that simple to get working for a pack update, and this suggestion would in theory be much easier for the end users if maybe harder for the pack maker, and as for configs this isnt so much about the actual merging or whatever and more about just getting the correct files to the right places on the user end, more for mod configs than minecraft settings; having defaultoptions does nothing if distributing updated defaults to put in there is going to be manual anyway

**Comment by Xetaxheb** [comment]
what part of the implementation do you think wouldn't be the same manual steps as creating a batch file?
you would still need to manually direct what files are overwritten and what files keep modified settings and what files are removed...

even if you're comparing Zip-A to Zip-B that still doesn't allow for modified client configs especially if the Zip-B had made changes themselves; and jar's are super easy to do in a batch file for the "remove outdated, add updated, add new" parts

i'll also direct you to "clip" 
especially in the context of "dir /b" and file tree comparing  ("   dir /b | clip   ")  

WinMerge is also especially helpful

**Comment by Hexality** [comment]
complementing this, a bit out of this context but also on the update side:
> An option to update the technic modpack right away instead of creating a new instance.

**Comment by Xetaxheb** [comment]
> complementing this, a bit out of this context but also on the update side:
> 
> > An option to update the technic modpack right away instead of creating a new instance.

feel free to elaborate how you think this should function on a technical level and in what ways this would be different from just creating your own diff+batch custom tailored to your own pack

you can't just throw a new zip at users instances and expect things to work

**Comment by Hexality** [comment]
just like technic launcher works, it basically checks what changed from a pack to the other and updates what changed.
I know this would be a bit complicated to you guys to make (unless the way that the thing is integrated changes), but the feature would basically be that, you get a discrete notification where it says the pack has updates, you click the button to update, then it gives the user a warning about possible problems with the update and then it updates what its needed.

**Comment by Hexality** [comment]
https://user-images.githubusercontent.com/17398632/127224542-aa1f9613-dc97-4aa5-b569-440dd5d6969b.mp4

**Comment by Xetaxheb** [comment]
TechnicLauncher has two separate "modes"

- one where packs are just hotlinked and keep a "what version of PACK does client have installed" 
- and one with their TechnicSolder api/database where it caches mods and md5's and mod lists for the modpack... a man in the middle I doubt multimc will implement or host (or rather is essentially curseforge/etc packs system anyway... just more json+hotlinks and less zip+metadata)

The latter is piecemeal like that but isn't relevant because of its nature, and the former doesn't resolve the existing issues with merging into/updating an existing install... namely that you either have to have an identified set of (or make guesswork about) what files to change.

Assuming you cached some metadata from the pack so you know what files the install ended up with after setup... (which leads to another issue i'll touch on in a minute)

Jars: sure, easy compare between list A and B and removing the ones that no longer exist on B and add the ones that are new on B.

Any other type of config files, scripts, anything other that may be modified or generated by the user though, there are numerous cases: 

-  how do you know it doesn't need regenerating/clearing/deleting? (i.e. "I change state from install during playing and should have been modified by new install but wasn't" bookmarks.ini from jei potentially)
-  how do you know it shouldn't be touched? (i.e. "I have player customized settings" bookmarks.ini from jei... again)
-  how do you know when a file that could be modified by the update (different md5) but shouldn't be overwriting a users changes? (still bookmarks.ini from jei in various cases... lmao)

All of this would require a diff patch or other intelligent setup on the modpack files end that just doesn't exist besides creating it manually or making a huge list of arbitrary files that the multimc team thinks is appropriate.... and will be wrong in many use cases.

Now, touching on the issue I said i'd get back to.... how would you even track this across multiple versions? User A installs packs version 1.0, updates to 1.1, then to 1.2; and User B installs packs versions 1.0, and updates to 1.2..... suppose 1.0>1.1 had a bugged file addition that breaks if the file isn't regenerated properly? 1.0>1.2 doesn't have the issue if you take a stance of "delete files that don't match/exist in the new metadata" but then you might end up deleting player configs/etc


**TL;DR the only ways of accomplishing an update to a modpack are:** 
1. preserving user data with a diff patch, which don't exist
2. doing a clean install, which means there's no point to the update button
3. removing old jars and adding new ones and overwriting existing files, which means potential compat/config issues (especially for files that need regenerating/removing and may not even have existed until generated in-game on-run)
4. having a custom GIT esque setup like TechnicSolder, which requires hosting/manual setup outside the major pack hubs (also has several of the issues mentioned anyway)


So unless someone is specifically requesting one of these broken features and thinks they'll get it through please anyone who wants to come up with a detailed technical writeup of a glorious solution feel free to chime in.

The only good solution I can think of is a tool for modpack creators to generate a set of diff patches quickly and easily between the versions of their packs and include it in some kind of /diffupdate/ folder for application with another tool that multimc could then integrate support for.

**Comment by spannerman79** [comment]
> The only good solution I can think of is a tool for modpack creators to generate a set of diff patches quickly and easily between the versions of their packs and include it in some kind of /diffupdate/ folder for application with another tool that multimc could then integrate support for.

**This, 100%!**

Idealy it should use Git as the backend.  Now as most modpack makers aren't exactly tech savvy a simple GUI frontend for Git would be ideal.  Most likley wirtten in Qt so it is cross platform.

The modpack version would be an annotated tag (example).

_But_, there is always a but, going towards a git based system _might_ cause issues with some mod devs as they want their curse points and only allow their mod to be downloaded directly from curseforge (yes those mod devs exist).

**Comment by Xetaxheb** [comment]
> > The only good solution I can think of is a tool for modpack creators to generate a set of diff patches quickly and easily between the versions of their packs and include it in some kind of /diffupdate/ folder for application with another tool that multimc could then integrate support for.
> 
> **This, 100%!**
> 
> Idealy it should use Git as the backend - [actual git](https://git-scm.com/), not some half baked rip off. Now as most modpack makers aren't exactly tech savvy a simple GUI frontend for Git would be ideal. Most likley wirtten in Qt so it is cross platform.
> 
> The modpack version would be an annotated tag (example).
> 
> _But_, there is always a but, going towards a git based system _might_ cause issues with some mod devs as they want their curse points and only allow their mod to be downloaded directly from curseforge (yes those mod devs exist).

Sorry, it sounds like you've misunderstood...? At no point am I advocating for bypassing mod distribution copyrights... I'm talking about modpack creators putting a set of basic diff patches in a subfolder inside their modpack zips that then can be applied via third party tools or multimc integration.  There would be no alternative hosting needed, it would still be entirely curseforge based but require modpack authors to take a few extra steps (and preferably do a manual review of their changes) when making an update to their packs.

**Comment by peterix** [comment]
You are severely overcomplicating this

**Comment by spannerman79** [comment]
No @Xetaxheb you have misunderstood.

>or making a huge list of arbitrary files that the multimc team thinks is appropriate.... and will be wrong in many use cases.

A proper version control system would be the only system that I could say that would be acceptable.  It is a known standard and _well_ used world wide.

> User A installs packs version 1.0, updates to 1.1, then to 1.2; and User B installs packs versions 1.0, and updates to 1.2..... suppose 1.0>1.1 had a bugged file addition that breaks if the file isn't regenerated properly? 1.0>1.2 doesn't have the issue if you take a stance of "delete files that don't match/exist in the new metadata" but then you might end up deleting player configs/etc

Git already handles this.  Don't need to reinvent the wheel here.

`git checkout v1.2` solves that above issue for user B.

But to handle this part specifically;

>"delete files that don't match/exist in the new metadata" but then you might end up deleting player configs/etc

Correctly annotated `.gitignore` handles this on first initial pack creation (ie v0.01), you include player specific data like `options.txt`, etc in it.

> while updating existing text files that would otherwise be overwritten; using diff patches.

**Literally** what a git patch entails.  An example of one such patch that was just recently merged with this repo
https://patch-diff.githubusercontent.com/raw/MultiMC/MultiMC5/pull/3537.patch

**Comment by Hexality** [comment]
>  mo0ComRbKg.mp4

also, about this, this isn't only for solder enabled modpacks but for all modpacks on technicpack.

**Comment by peterix** [comment]
The suggestion makes sense, the technical details in the discussion not really.

If you want to use git now, you can already. Just use it. `git init` your instances. Go crazy with it.

But absolutely no modpack platform of any note uses that. It's not a solution for 99% of the users.

Import 'over' an instance sounds like a decent idea.

Take new pack and its content (C), add keeping track of what was originally imported (A), combine with the current state (B)...
Then you can diff (A) to (B) to determine what the user changed, and use that to inform the transform from (A) to (C).

When you don't have (A), you can't make as good of a job going from (B) to (C), but a some heuristics should help there...

Don't mix this with git. You can already use git and it needs absolutely no integration with the launcher. Actually, standalong git tools are most likely always going to be better even IF there is any git integration. And git does not help 99% of the users with modpacks from platforms that do not, and have never used git.

**Comment by Hexality** [comment]
> The suggestion makes sense, the technical details in the discussion not really.
> 
> If you want to use git now, you can already. Just use it. `git init` your instances. Go crazy with it.
> 
> But absolutely no modpack platform of any note uses that. It's not a solution for 99% of the users.
> 
> Import 'over' an instance sounds like a decent idea.
> 
> Take new pack and its content (C), add keeping track of what was originally imported (A), combine with the current state (B)...
> Then you can diff (A) to (B) to determine what the user changed, and use that to inform the transform from (B) to (C).
> 
> When you don't have (A), you can't make as good of a job going from (B) to (C), but some heuristics should help there...
> 
> Don't mix this with git. You can already use git and it needs absolutely no integration with the launcher. Actually, standalone git tools are most likely always going to be better even IF there is any git integration. And git does not help 99% of the users with modpacks from platforms that do not, and have never used git.

that's my original suggestion, the person who overcomplicated it was xetaxheb :p
